src/source/custom_source.js

```diff
@@ -0,0 +1,336 @@
+    dispatcher: Dispatcher;
+    _loaded: boolean;
+
+    constructor(id: string, implementation: CustomSourceInterface<T>, dispatcher: Dispatcher, eventedParent: Evented) {
+        super();
+        this.id = id;
+        this.type = 'custom';
+        this.dataType = 'raster';
+        this.dispatcher = dispatcher;
+        this.setEventedParent(eventedParent);
+        this.implementation = implementation;
+
+        this.scheme = 'xyz';
+        this.minzoom = 0;
+        this.maxzoom = 22;
+        this.tileSize = 512;
+
+        this._loaded = false;
+        this.roundZoom = true;
+
+        if (!implementation) {
+            this.fire(new ErrorEvent(new Error(`Missing implementation for ${this.id} custom source`)));
+        }
+
+        // $FlowFixMe[prop-missing]
+        implementation.update = this.update.bind(this);
+
+        // $FlowFixMe[prop-missing]
+        implementation.coveringTiles = this.coveringTiles.bind(this);
+
+        extend(this, pick(implementation, ['dataType', 'scheme', 'minzoom', 'maxzoom', 'tileSize', 'attribution', 'minTileCacheSize', 'maxTileCacheSize']));
+    }
+
+    serialize(): Source {
+        return pick(this, ['type', 'scheme', 'minzoom', 'maxzoom', 'tileSize', 'attribution']);
+    }
+
+    load() {
+        this._loaded = true;
+        this.fire(new Event('data', {dataType: 'source', sourceDataType: 'metadata'}));
+        this.fire(new Event('data', {dataType: 'source', sourceDataType: 'content'}));
+    }
+
+    loaded(): boolean {
+        return this._loaded;
+    }
+
+    onAdd(map: Map): void {
+        this.map = map;
+        if (this.implementation.onAdd) {
+            this.implementation.onAdd(map, this.load);
+        } else {
+            this.load();
+        }
+    }
+
+    onRemove(map: Map): void {
+        if (this.implementation.onRemove) {
+            this.implementation.onRemove(map);
+        }
+    }
+
+    loadTile(tile: Tile, callback: Callback<void>) {
+        const controller = new AbortController();
+        const signal = controller.signal;
+
+        // $FlowFixMe[prop-missing]
+        tile.request = this.implementation.loadTile(tile.tileID.canonical, {signal})
+            .then(tileLoaded.bind(this))
+            .catch(error => {
+                // silence AbortError and 404 errors
+                if (error.code === 20 || error.code === 404) return;
+                callback(error);
+            });
+
+        // $FlowFixMe[prop-missing]
+        tile.request.cancel = () => controller.abort();
+
+        function tileLoaded(data) {
+            delete tile.request;
+
+            if (tile.aborted) return callback(null);
+            if (!data) return callback(null);
+            if (!isRaster(data)) return callback(new Error(`Can't infer data type for ${this.id}, only raster data supported at the moment`));
+
+            this.loadTileData(tile, data);
+            tile.state = 'loaded';
+
+            cacheEntryPossiblyAdded(this.dispatcher);
```

Comment:
I think this can be removed since these custom tiles never add to our cache, right?